ui/app/routes/vault/cluster/secrets/backend/secret-edit.js

```diff
@@ -98,53 +97,93 @@ export default Route.extend(UnloadModelRoute, {
-          //don't have access to the metadata, so we'll make
-          //a stub metadata model and try to load the version
-          if (modelType === 'secret-v2' && err.httpStatus === 403) {
-            let secretModel = this.store.createRecord('secret-v2');
-            secretModel.setProperties({
-              engine: backendModel,
-              id: secret,
-              // so we know it's a stub model and won't be saving it
-              // because we don't have access to that endpoint
-              isStub: true,
-            });
-            let targetVersion = params.version ? parseInt(params.version, 10) : null;
-            let versionId = targetVersion ? [backend, secret, targetVersion] : [backend, secret];
-            return this.store
-              .findRecord('secret-v2-version', JSON.stringify(versionId), { reload: true })
-              .then(versionModel => {
-                secretModel.set('selectedVersion', versionModel);
-                return secretModel;
-              });
-          }
+    let secretModel;
+
+    let capabilities = this.capabilities(secret);
+    try {
+      secretModel = await this.store.queryRecord(modelType, { id: secret, backend });
+    } catch (err) {
+      if (err.httpStatus === 403 && (modelType === 'secret-v2' || modelType === 'secret')) {
+        await capabilities;
+        // can't read the path and don't have update permission, so re-throw
+        if (!capabilities.get('canUpdate') && modelType === 'secret') {
           throw err;
-        }),
-      capabilities: this.capabilities(secret),
-    });
+        }
+        // don't have access to the metadata for v2 or the secret for v1,
+        // so we make a stub model
+        secretModel = this.store.createRecord(modelType);
+        secretModel.setProperties({
+          id: secret,
+          // so we know it's a stub model and won't be saving it
+          // because we don't have access to that endpoint
+          failedServerRead: true,
+        });
+      } else {
+        throw err;
+      }
+    }
+    await capabilities;
+    // after the the base model fetch, kv-v2 has a second associated
+    // version model that contains the secret data
+    if (modelType === 'secret-v2') {
+      let targetVersion;
+      if (secretModel.currentVersion) {
+        // we have the secret metadata, so we can read the currentVersion but give priority to any
+        // version passed in via the url
+        targetVersion = parseInt(params.version || secretModel.currentVersion, 10);
+      } else {
+        // we've got a stub model because we can't read the metadata
+        targetVersion = params.version ? parseInt(params.version, 10) : null;
+      }
+
+      let versionId = targetVersion ? [backend, secret, targetVersion] : [backend, secret];
+
+      // if we have the metadata, a list of versions are part of the payload
+      let version = secretModel.versions && secretModel.versions.findBy('version', targetVersion);
```

Comment:
hmm, so i've noticed after reading through this method a few times that this is the point where i start losing track of what's happening. i know there's a lot of logic that needs to happen inside `model`, but i wonder if we could break some of it up either into smaller methods like `getSecretModel` `getTargetVersion` `fetchVersion` etc for easier legibility. if it's too difficult to set that up, i think adding comment headers above each code block would be a good alternative, especially since the existing comments you have here are so helpful. whaddya think?

Reply:
Yep, that's a great idea! Have an update to push 🔜. Thanks!